src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImplicitUsingsValueProvider.cs

```diff
@@ -0,0 +1,200 @@
+                    {
+                        project.RemoveItems(
+                            project.GetItems("Using").Where(i => string.Equals(existingUsing.Include, i.EvaluatedInclude, StringComparisons.ItemNames) && !i.IsImported)
+                        );
+                    });
+
+                }
+                catch (Exception ex)
+                {
+                    // if an import comes from a targets file, or else if there's a race condition we can't remove it. otherwise throw
+                    if (ex is not ArgumentException && ex is not InvalidOperationException)
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        foreach (ImplicitUsing existingUsing in existingUsings)
+        {
+            if (usingsToSet.Any(usingToSet => usingToSet.Equals(existingUsing)))
+            {
+                usingsToSet.Remove(existingUsing);
+            }
+        }
+        
+        // Verify we have at least one valid import to add before acquiring a write lock.
+        if (usingsToSet.Any(importToAdd => importToAdd.Include.Length > 0))
+        {
+            await _projectAccessor.OpenProjectForWriteAsync(_configuredProject, project =>
+            {
+                foreach (ImplicitUsing usingToSet in usingsToSet)
+                {
+
+                    List<ImplicitUsing> existingUsingsOfIncludeToRemove = existingUsings.FindAll(existingUsing =>
+                        string.Equals(usingToSet.Include, existingUsing.Include, StringComparison.Ordinal) && !existingUsing.IsReadOnly && !usingToSet.Equals(existingUsing));
+
+                    if (existingUsingsOfIncludeToRemove.Count > 0)
+                    {
+                        project.RemoveItems(
+                            project.GetItems("Using")
+                                .Where(i => string.Equals(usingToSet.Include, i.EvaluatedInclude, StringComparisons.ItemNames) && !i.IsImported)
+                        );
+                    }
+                    
+                    if (usingToSet.Include.Length > 0 && (usingToSet.Alias is null || usingToSet.Alias.Length > 0))
+                    {
+                        var usingMetadata = new List<KeyValuePair<string, string>>();
+                        if (usingToSet.Alias is { } alias)
+                        {
+                            usingMetadata.Add(new KeyValuePair<string, string>("Alias", alias));
+                        }
+
+                        if (usingToSet.IsStatic)
+                        {
+                            usingMetadata.Add(new KeyValuePair<string, string>("Static", usingToSet.IsStatic.ToString()));
+                        }
+                        
+                        project.AddItem("Using", usingToSet.Include, usingMetadata);
+                    }
+                }
+            });
+        }
+
+        return null;
+    }
+
+    internal class ImplicitUsing
```

Comment:
Could this class be promoted to be its own file? It might help specially when adding the VB experience, I suspect we would re-use this.

Reply:
Bonus points for turning this into a C# `record`. ;-)

Reply:
We target <net5 in this project, so we run into the following error: The predefined type 'System.Runtime.CompilerServices.IsExternalInit' must be defined or imported in order to declare init-only setter

Unfortunately, we are not able to use record types without a hacky workaround.